rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,669 @@
+  operations).
+
+This latter directive can be initially pessimistically implemented as a synonym
+of `#[cfg(target_has_atomic_load_store = N)]`, then gradually extended to
+support targets which have no or non-native support of `Relaxed` atomics
+but do have native load/store instructions of a certain width, such as `nvptx`.
+
+In this way, the proposed volatile atomic operation API can largely re-use the
+already existing atomic operation support infrastructure, which will greatly
+reduce effort duplication between these two closely related functionalities.
+
+---
+
+This RFC currently proposes to expose the full power of LLVM's atomic volatile
+operations, including e.g. read-modify-write operations like compare-and-swap,
+because it is consistent with the atomics operation API and could have
+legitimate uses in interprocess communication scenarios, as a marker of the
+nuance between well-optimized program-local synchronization and worst-case FFI
+synchronization. See Unresolved Questions section for more details.
+
+However, the fact that these operations do not necessarily compile into a single
+hardware instruction is arguably a footgun for volatile's use cases, and it
+could be argued that initially only stabilizing loads, stores and `Relaxed`
+atomic ordering would be more prudent. This will be revisited in the
+alternatives section of this RFC.
+
+---
+
+As currently designed, this RFC uses `arbitrary_safe_types` to give method-like
+semantics to a `NonNull` raw pointer. This seems necessary to get reasonable
+ergonomics with an atomics-like wrapper type approach. However, it could also be
+argued that a `VolatileXyz::store(ptr, data, ordering)` style of API would work
+well enough, and avoid coupling with unstable features. Similarly, the use of
+`NonNull` itself could be debated. This will be revisited in the alternatives
+section of the RFC as well.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Deprecating the existing volatile operations will cause language churn. Not
+deprecating them will keep around two different and subtly incompatible ways to
+do the same thing, which is equally bad. It could be argued that the issues with
+the existing volatile operations, while real, do not warrant the full complexity
+of this RFC.
+
+As mentioned above, exposing more than loads and stores, and non-`Relaxed`
+atomic memory orderings, also muddies the "a volatile op should compile into one
+hardware memory instruction" narrative that is so convenient for loads and
+stores. Further, compatibility of non-load/store atomics in IPC scenario may
+require some ABI agreement on how atomics should be implemented between the
+interacting programs.
+
+If this is perceived to be a problem, we could decide to do away with some of
+the complexity by initially focusing on a restricted subset of this proposal
+which only supports `Relaxed` loads and stores, and saving more complex atomic
+operations as a future extension.
```

Comment:
I am strongly in favor of this. I think if people want CAS or want to control the ordering, they should use `AtomicX` types, and if they *also* need the same accesses to be volatile, then the `AtomicX` types should provide volatile operations. Having to duplicate all these `fetch_*` methods seems silly.

Reply:
I'll reword in this direction and mark atomic volatile as a possible future extension, then.

Reply:
Done.